### PR TITLE
[NEW] Message retention policy and pruning

### LINKED ISSUE #1861: Ability to delete multiple messages at once
Hi all,

I suggest to add a button for clean all (or partial?) messages in group : There is not necessarily need to archive some discussions

Thomas

### LINKED ISSUE #2355: Ability to delete chat history
Your Rocket.Chat version: 0.19.0

This is not an issue per se, but more of an enhancement request if possible.

> It would be nice to have the ability to delete a chat's history, could be a room's history, or a 1 to 1 chat history. Totally, all messages displayed in a given room. Or delete all messages older than $DATE , or delete all messages between DATE1 and DATE2.

### LINKED ISSUE #2639: Add ability to delete channel's history/media
Your Rocket.Chat version: 0.23.0

> This is not an issue but a feature request "nice to have"

**1.** The ability to delete a channel's or 1 to 1 conversation history and media shared.

**2.** For small servers, the ability to automatically delete/overwrite the memory used for conversation and shared media history. Say, if it's older than X days, months. **OR**, whenever the conversation history or media buffer is over X amount of gigabytes, then start overwriting the oldest line/media.

### LINKED ISSUE #2700: Deleting user doesn't delete messages
0.21.0, commit hash ...dac592
1. Log in as user X and leave some messages in #general
2. In a different tab, log in as an admin and delete user X. Notice the warning dialog saying that the user's messages will be deleted.

A third user who was watching #general reports that user X's messages are still there. Hard reloading the page no longer shows the deleted user's messages.
When I wanted to join #general as the admin user, the preview mode still showed the messages.

Possibly related: #1149, #739

### LINKED ISSUE #6193: Delete all Attachments / Files from a Channel /Chat
Rocket.Chat Version:  0.52.0
Running Instances:  1
DB Replicaset OpLog: disabled
Node Version: V4.5.0

Hi there, 

We would recoment a function wich eareses all Attachments in one Channel or Chat. 

Example 1: a private chat between two Users got now more than 100 files uplaoded. 
Example 2: a channel with more than 30 Users and a lot of files has been added to that channel. 

Many of that files are no Longer used, but its hard work to delet them all single handed.

### LINKED ISSUE #6749: Deleting old messages?
Rocket.Chat Version: 0.55.0 rc-6
Running Instances: 1
DB Replicaset OpLog:
Node Version: 4.5.0

Hi

Is there a way to delete (prune) old messages automatically?  Ideally i would only want something like the last 50 messages posted to the room to be saved and viewable for any particular user.

I am rather new to Rocket.chat and did not find this function.   It seems to store the entire history since the beginning and that would take up a lot of disk space if left on long enough.

Thanks

### LINKED ISSUE #7228: Deleting Chat Room Logs +30 day crontab
I've looked to see if anyone has come up with a solution for this.
I would like to be able to run a crontab job which deletes any messages from any room which are older than 30 days.  Is this at all possible?  My concern is that the Mongo database will quickly fill with messages and cause the server to run out of disk space on active channels.  I've seen posts about using the API but it isn't clearly explained how to achieve this.  Our Rocket Chat is running in a Snap environment.
Thanks

Rocket.Chat Version: 0.56.0
Running Instances: 1
DB Replicaset OpLog: Enabled
Node Version: 4.8.1

### LINKED ISSUE #8321: System wide auto delete feature
Would like to see a system wide setting available to owner/admin that would set a "drop dead date" / auto delete period for all message on the system. Public channels, private groups & private messages. Owner/admin could set entire environment to auto delete in 7, 30, etc... days.

I believe chat should be ephemeral like real conversations. I believe therefore that logs of chat should have the option of simply going away if the owner/admin wants to promote that kind of environment.

Cheers! Rich G

### LINKED ISSUE #8757: [Feature request] Per channel option to clean up messages after selected period of time.
Channel option to clean up the messages after a selected period of time would be nice. Room owners would be able to set the message retention time.

The main purpose of this would be additional security (in case of database leak), but also preventing database growth.

Similar option is requested here #8749, but since the private channels contain membership information, setting the rooms up from the scratch would cause unnecessary work.

### LINKED ISSUE #9317: Automatically delete attached files
I’d like to recommend implementing an option regarding file upload, that allows all attachments to be deleted after a certain amount of time. For example: every file that is older than 14 or 30 days would be deleted from all channels on the server. Furthermore, it would be advantagous to have an option that allows a maximum number of uploaded files, and if the users upload more files than allowed, the oldest file will be deleted. Rright now all uploaded files are stored on the server, it takes up a lot of space in the long run. Furthermore, attachments from private conversations cannot be deleted, so some users can take up GBs of space. Thank you very much.

### LINKED ISSUE #9374: Deleting a user should instantly remove messages from that user
### Description:

If I delete a user, the messages are removed, but this isn't reflected until clients reload the channel(s). 

### Server Setup Information:

* Version of Rocket.Chat Server: 0.60.3
* Operating System: Ubuntu 16.04
* Deployment Method(snap/docker/tar/etc): docker
* Number of Running Instances: 1
* DB Replicaset Oplog: n/a
* Node Version: 8

### Steps to Reproduce:

1. Create extra user
2. Type some messages in a channel
3. Delete extra user

### Expected behavior:

I would expect all messages to be instantly removed but they are not. 

### Actual behavior:

Messages remain until the client is reloaded.

### LINKED ISSUE #10870: API cleanRoomHistory Does Not Remove Attached Media
The api which uses the `cleanRoomHistory` method does not clean up the attached files/media.

### LINKED ISSUE #11299: Deleting room doesn't delete files
When deleting a room, the files within the room are not deleted and remain on the mongo DB.

### LINKED ISSUE #11468: Deleting user doesn't delete their files
When a user is deleted, their files stay on the DB.

### PR DESCRIPTION
**Pull request by vynmera**
Closes #6749
Closes #8321
Closes #9374
Closes #2700
Closes #2639
Closes #2355 
Closes #1861
Closes #8757
Closes #7228
Closes #10870
Closes #6193 
Closes #11299
Closes #11468
Closes #9317
Closes #11300 (will incorporate a fix to this PR's issue)
Closes #11046 (will incorporate a fix to this PR's issue)
Contributes to #5944 
Contributes to #11475
_...and possibly more!_

This PR makes deleting messages (automatically and manually) a lot easier on Rocket.Chat.

- [X] Implement a bulk message deletion notification, to quickly push large message deletions to users without reload
  - [X] Use it in `rooms.cleanHistory`
  - [X] Use it in user deletions
- [X] Completely remove cleanChannelHistory as required by v0.67
  - [X] Remove server method `cleanChannelHistory`
  - [X] Remove REST API `channels.cleanHistory`
- [x] Implement a sidebar option to clean history
  - [x] Basic implementation
  - [x] Allow excluding pinned messages
  - [x] Allow attachment-only mode
  - [x] Allow specifying user(s) to narrow down to
    - [x] Also update REST API
    - [x] Also update docs
  - [x] Break the deletion into multiple different requests, so the client can keep track of progress
  - [x] Clear animation / progress bar for deleting
- [x] Retention policy
  - [X] Global, set by admin
    - [X] Global timer that runs every second and deletes messages over the set limit
      - [X] Can change its timer's resolution to prevent insane CPU overhead
    - [X] Admin can decide what room types to target (channels, groups and/or DMs)
    - [X] Allow excluding pinned messages
    - [X] Allow attachment-only mode
  - [x] Per-channel, set by those with a new permission
    - [x] Disabled when master switch off
    - [x] Set in channel info
    - [x] Can override global policy with a switch that requires `edit-privileged-setting`
    - [x] Allow excluding pinned messages
    - [x] Allow attachment-only mode
    - [x] Uses same global timer for cleanup
  - [X] Message at start of channel history / in channel info if there is a retention policy set
  - [x] Message in channel info if there is a retention policy set on that channel specifically
- [X] Make cleaning history also delete files (completely!)
  - [X] Manual purging
  - [X] Automatic purging
- [x] Make other deletions also delete files
  - [x] User deletion
    - [X] Own messages
    - [x] DMs with them's partner messages
  - [x] Room deletion
- [x] Cleanup
- [x] Finish related [docs](https://github.com/RocketChat/docs/pull/815)
- [x] Link to the docs in the settings

Please suggest any cool changes/additions! Any support is greatly appreciated.

**Breaking change:** This PR removes REST API endpoint `channels.cleanHistory` and Meteor callable `cleanChannelHistory` as per the protocol specified for them.

bzzzzzzzz

### CODE REVIEW

**Comment by vynmera** [comment]
A simple implementation for the flexbar is here! :tada:

**Comment by vynmera** [comment]
The flexbar now has a progress indicator, and sends the request in small pieces (it uses a new `limit` parameter), to keep track! :tada:

**Comment by vynmera** [comment]
A simple implementation of a global retention policy now exists! :tada: It's configurable in Administration.



The example above looks like this in chat:

**Comment by wdimd** [comment]
This is awesome, thank you so much!!!

**Comment by rodrigok** [comment]
Wow, awesome work @vynmera 
Can't wait to review this PR 😄

**Comment by TheMythPT** [comment]
This is top top top!
Great work @vynmera 
:tada:

**Comment by hExPY** [comment]
Aaawesome!!!!

**Comment by mddvul22** [comment]
Is it possible to have different maximum message age values for direct messages, private groups, and public channels?  It would be nice to have different values for those three things.

**Comment by vynmera** [comment]
@mddvul22 That's a good idea, I'll be on it in a moment! I can imagine people needing to keep private groups on a shorter timer.

**Comment by vynmera** [comment]
@mddvul22 Just pushed this:

**Comment by mddvul22** [comment]
Awesome!

**Comment by TheMythPT** [comment]
just a small idea. as text messages don't take that much space unless you have tons of users, would be nice to have an option just to delete older uploaded files (avatars should be kept safe).
it could be a good addition perhaps?

**Comment by himpich** [comment]
@vynmera 
Suggestion:
- include / exclude [via options] related attachments
- include / exclude [via options] sticky posts

otherwise: Thanks for taking care of this long missed feature!

**Comment by vynmera** [comment]
@TheMythPT That should be possible! Just a simple `[X] Files only` in settings.

@himpich Excluding pinned posts is a good idea. As for excluding attachments, the issue with that is that it harms the database integrity by creating files not connected with any message. In the way the code currently works, that means that if you at any point _do_ want to delete them, it is very difficult (will take a lot of time) to do that - the current implementation of `rooms.cleanHistory` causes this issue as well, and we ended up just deleting all files beyond a certain age using mongo shell :(

**Comment by vynmera** [comment]
As per @TheMythPT and @himpich 's suggestions:

You can now set whether to remove pinned messages, and whether to only remove attachments in both manual and autoprune! :tada:

**Comment by vynmera** [comment]
@eszeus You can clone my fork `git clone https://github.com/vynmera/Rocket.Chat.git`, go to it `cd Rocket.Chat/`, then go to my branch `git checkout purge`, and deploy it with `meteor npm start`. For more info on how to deploy a development version, check this out: https://rocket.chat/docs/developer-guides/quick-start/

**Note that it's a WIP build, don't expect everything to work correctly (or at all). I _highly_ advise you to not use it in production :)**

When I update anything, you can `git pull` to download the newest version.

Once I have per-channel retention policies in, perhaps @engelgabriel can deploy it to Heroku so everyone can fiddle with it?

**Comment by mddvul22** [comment]
This is all great work!  Thanks for working on this!  Out of curiosity, would it be difficult to add something like a default message age value, but then allow admins to override that value for a specified public channel, or private room?  Also, will this have elaborate permissions controls, or will these message age limit controls only be configurable by the admins? (admin-only seems preferable to me)

**Comment by vynmera** [comment]
@mddvul22 As I'm coding it right now, **admins** can set a global retention policy across all channels.
**Channel owners** (those with `edit-room`) can add _additional, stricter_ policies onto their channels. So if the global policy is set to 30 days, a channel owner can lower it to 7 days, but not extend it to 60. Or if the global policy deletes messages _and_ files, the channel owner can't change it to only files.
**Admins** (those with `edit-privileged-setting`) can, however, toggle a switch that bypasses this restriction and allows to extend or even remove the policy from a channel.

See this screenshot:


Is this OK with you?

**Comment by mddvul22** [comment]
I think that should work for us. Thanks!

**Comment by kaiiiiiiiii** [comment]
Will there be an option to disable the hole feature (manual deletion)? I really like it, but no one on our company chat should be able to delete any messages as they were created during work time and belong to our company :)

**Comment by vynmera** [comment]
@kaiiiiiiiii The manual deletion is already hooked up to the permission `clean-channel-history`, so you can simply remove that permission from everyone but admins.
Editing the retention policy is currently set to `edit-room`, perhaps it would be better to break this out into a new permission, perhaps `edit-room-retention-policy`? Then we can remove that permission from channel owners too.

Should we put a migration to change the permissions to allow channel owners access to `clean-channel-history`, or should admins decide first whether this is a feature they wish to activate for channel owners to use? In the same train of thought, should the default permissions for `edit-room-retention-policy` be to admins only, or to channel owners too?

**Comment by vynmera** [comment]
Let me rephrase that, what should the default permissions be?

Prune channel -> admin only or for channel owners too?
Same with setting per-channel retention policies - admin-only or allow channel owners to set them too?

Defaults, of course - can be edited in Permissions

**Comment by kaiiiiiiiii** [comment]
I’m always a fan of keeping things the way they are and let admins decide if they want to activate something or not :)

So I would prefer admin-only as a default setting.

Cheers,
Kai

**Comment by TheMythPT** [comment]
**Prune channel -> admin only or for channel owners too?**
On this I'm for channel owners to be able to prune the channel, if there's a retention policy enabled by admins

**Same with setting per-channel retention policies -> admin-only or allow channel owners to set them too?**
Once again channel owners should be able to set a channel policy retention if _lower_ than the global one set by the admin.

**Comment by vynmera** [comment]
Channels can now have custom policies! :tada: 

Those with the new `edit-room-retention-policy` permission can set policies onto channels. By default, this permission is given to **admins** only, but if you want to allow channel owners to set policies for their channels, grant them the permission and they'll be able to go right ahead.



By default, a per-room retention policy can only be set to be **stricter** than the global one. If there is a global policy for 30 days, a per-room policy can not extend it to 60, but _can_ shorten it to 7. Furthermore, if the global policy is set to not exclude pinned messages, switching "Exclude pinned" will have no effect (same goes for the file-only mode).
This can be circumvented by switching "Override global retention policy", but that switch requires the `edit-privileged-setting` permission. You can also use this to turn off the global policy for a channel - simply switch "Override global retention policy", and set "Automatically prune old messages" to **OFF**.

**Comment by vynmera** [comment]
Anything else I should add/edit? :)

**Comment by TheMythPT** [comment]
LGTM. If I find any useful ideas I'll let you know! :)

A Qui, 28 de jun de 2018, 16:37, フィンメーラ <notifications@github.com> escreveu:

**Comment by vynmera** [comment]
@wdimd WIP indeed means it's not yet ready for production and is still in development. Non-bug fixes will be pushed to full releases, and since 0.66.0 just went live, this feature will probably go to 0.67.0. That version is scheduled to launch **July 27th**. So about a month of waiting :)
Once I'm done and I take this PR out of [WIP] status, and the changes get approved, you can run the commands I describe above [here](https://github.com/RocketChat/Rocket.Chat/pull/11236#issuecomment-400582317) and expect a pretty solid experience, but I still advise against deploying it on production :smirk:

**Comment by vynmera** [comment]
You can now specify a precision for the timer, as running the purge script every second might be too processing power-expensive for large servers (that don't need high-precision purging) or might cause spam in the logs for those who set it to be verbose.

**Comment by vynmera** [comment]
The new fields `limit` `excludePinned` and `filesOnly` can now be used from the `rooms.cleanHistory` REST API.
I'll create a docs update in a moment.

**Comment by vynmera** [comment]
Companion docs update: https://github.com/RocketChat/docs/pull/815 (also WIP)

**Comment by maxdwit** [comment]
@vynmera could you please update to meet the base branch

**Comment by vynmera** [comment]
I am taking this out of [WIP] now because I feel all features I can think up for this are finished. Feel free to review it, request changes or new features though!

**Comment by vynmera** [comment]
I've made the settings more organized. Switching off the `Enable` switch now completely disables the feature - to allow per-channel policies without using any global policy, simply keep all the `Applies to` switches off.

**Comment by marzie82** [comment]
Hi,
that's a great idea- how can i activate this function?

**Comment by vynmera** [comment]
# Summary
This PR adds retention policies, cleaning channels from the UI, and some retention-related fixes to Rocket.Chat.

## :clock1030: Retention policies
You may now select how long you wish channels' messages and files to exist for!
There are settings for global and per-channel policies:


_Global policy_


_Per-channel policy_

Both of these will result in a nice header showing how your messages will be handled.



For more information on retention policies, the docs can be read [here](https://github.com/vynmera/docs/blob/b6c0881504f002cbf1d61efcbc9be042d9e19ce2/administrator-guides/retention-policies/README.md) or, when pushed, [here](https://rocket.chat/docs/administrator-guides/retention-policies/)!

## :boom: Channel pruning
What was before relatively difficult to do with the REST API is now possible straight from the client. You may delete any content from any channel, super quickly.




For more information on channel pruning, the docs can be read [here](https://github.com/vynmera/docs/blob/b6c0881504f002cbf1d61efcbc9be042d9e19ce2/user-guides/channel-actions/README.md) or, when pushed, [here](https://rocket.chat/docs/user-guides/channel-actions/)!

## :warning: Breaking changes
The REST API endpoint `channels.cleanHistory` and Meteor callable `cleanChannelHistory` have been removed.

## :heavy_plus_sign: Other changes
- The REST API endpoint `rooms.cleanHistory` has 4 new parameters.
- User deletions now show up instantly.
- User deletions now remove all the user's files too.
- User deletions now delete all the files in the rooms that will be deleted due to it.
- Room deletions now remove all the room's files too.
- Cleaning history now shows up instantly.
- Cleaning history now also deletes files according to the chosen parameters.
- Add/update a few methods

## :heart: Special thanks
@ggazzo for reviewing and helping with the code
The Rocket.Chat team for general support
The community for inspiration

**Comment by chatnl** [comment]
This is probably a stupid question. I just installed Rocket Chat with the Snap method a few days ago. I was looking for this new functionality to clear a rooms contents every night. This seems what I have been looking for. Can you please tell me how I can update my rocket chat in order to get these new options?

Thanks!

**Comment by vynmera** [comment]
@chatnl If you're using Snap, I'm afraid you'll have to wait for the release cycle to finish. 0.67.0 will be released around the 27th - Snap will autoupdate around then (might take a few days to test whether everything works before the team deploys the new Snap - they autoupdate after all).

You can test out the feature if you want though, a bit up this thread I described how to do so (clone the repo and run `meteor npm start`). But you can't really tinker with a Snap.

**Comment by chatnl** [comment]
@vynmera OK, well I didn't even know that snap autoupdates... Is there more information about how and when it does that? I am running a live chatbox and don't want to run into problems because of an update going wrong. If I know when these updates are going to happen I could schedule snapshots of the server and make sure that I can revert to a previous version.

Thanks for the quick reply!

**Comment by vynmera** [comment]
@chatnl Well, snaps check for updates every day, and the team pushes a new snap a few days after the release (27th of the month).

Here's some docs on using `snap`: https://docs.snapcraft.io/reference/snap-command#revert

For example, to undo an update and all changes since that update, you may `snap revert`.

**Comment by MarcosSpessatto** [comment]
@ggazzo you can close this PR. https://github.com/RocketChat/Rocket.Chat/pull/11046

**Comment by serviceman** [comment]
When is this expected to be added to a production release? I am working on a community site and I'd much rather wait for this feature before the release.

**Comment by ggazzo** [comment]
@serviceman next release `0.68` probably july 27th :)

**Comment by marzie82** [comment]
Great feature- many thanks !!!
Just one question- how do I get it? My Snap was updated two days ago- now I am on Version 0.66.3?
Any idea what's wrong?

**Comment by himpierre** [comment]
@marzie82 You need rocketchat 0.68.0.

**Comment by himpierre** [comment]
Ability to delete channels / dm automatically after x time with no message would be really nice.

**Comment by marzie82** [comment]
@himpierre Thanks ;) 
can you please let me know when SNAP 0.68.0 will be released?

**Comment by chatnl** [comment]
As far as I can see, SNAP 0.68.0 is not available yet. Not in the stable update channel and also not in the candidate channel. The latest version is still 0.66.3. I would also advice not to use updates from the candidate channel in a production environment. Better wait until there is a stable version. Maybe development can let us know when they expect it to be published as a stable version?

**Comment by vynmera** [comment]
@svennd This is already possible with the per-channel rules: for every channel, if you have the `edit-room-retention-policy` permission, you can change the retention policy on a channel-to-channel basis (no policy, different policy, etc..) :)

**Comment by svennd** [comment]
I apologize, I clearly missed where to find this feature, super ! Perhaps other people not finding it : 
pruning_per_channel

**Comment by claures** [comment]
Hello,
Great future does this also delete the file attachments?
-BR

**Comment by CH1PSET1** [comment]
Hello. Tell me, I have a version 0.68.3 server. Why does my automatic cleaning function not work for private messages? With what it can be connected? Global settings enable cleaning for channels, rooms and private messages. But it cleans only in channels and groups. In personal messages from the top it is written, Messages older than 14 days are automatically cropped here, but messages remain for an older period. In manual, if clean, it cleans.

And the second question. In the channel, in groups and private messages, if I choose to clear all messages, they are cleared in the web and desktop versions of the application, and in the mobile application the story remains, with what it too can be connected? And if I delete messages individually, they are deleted everywhere. Thank you.

**Comment by vynmera** [comment]
@CH1PSET1 As for the messages not being cleared automatically, there may be not enough activity in the channel: https://github.com/RocketChat/Rocket.Chat/pull/11725#issuecomment-411789252 discusses this further. (I've started some tests in the background and will return with the results once I've tested some more, but this might be the cause of the issue)

As for prunes not showing up in the mobile apps, this PR implements a new type of Notification. So the mobile apps will have to be updated to reflect this change. The messages _are_ gone, but the phone app doesn't know this because it not yet handles these new notifications. Once the mobile app is updated, this will work.

**Comment by maxdwit** [comment]
@vynmera thanks for bringing this into rocket.chat.

**Comment by dhpowrhost** [comment]
Will this be implemented in livechat also?

**Comment by GoetheG** [comment]
Hi guys. I wish I had the chance to decide for every channe whether a user can see historic messages. That's important for clubs for example. A chat can get messy and sometimes to informal and new guys may have a different wording. It would come in handy if I could set specific settings for a user. For example if I want to exclude a new added user in a channel to see previous chats.

**Comment by GoetheG** [comment]
This is what I am referring to:
https://github.com/RocketChat/Rocket.Chat/issues/1442

**Comment by UPD780C-1** [comment]
I am not sure if anyone else is still here. But just for my personal requirements, I recently built a Rocketchat server. Due to some security reasons, I have been looking for the information how to purge chat message in a short time of period. Then I found this page!

I tried everything I could with my current skill set, but I am currently stuck...
After installing the required packages, I finally executed `meteor npm start` following this article and the link below.
https://rocket.chat/docs/developer-guides/quick-start/

However, its getting these errors, and the process's been aborted. 

`=> Started proxy.
Browserslist: caniuse-lite is outdated. Please run next command `npm update caniuse-lite browserslist`
WARNING: npm peer requirements (for juliancwirko:postcss) not installed:
 - postcss@7.0.17 installed, postcss@^6.0.22 needed
 - postcss-load-config@2.1.0 installed, postcss-load-config@^1.2.0 needed

Read more about installing npm peer dependencies:
  http://guide.meteor.com/using-packages.html#peer-npm-dependencies

   Minifying app stylesheet                  -
  Replace Autoprefixer browsers option to Browserslist config.
  Use browserslist key in package.json or .browserslistrc file.

  Using browsers option cause some error. Browserslist config
  can be used for Babel, Autoprefixer, postcss-normalize and other tools.

  If you really need to use option, rename it to overrideBrowserslist.

  Learn more at:
  https://github.com/browserslist/browserslist#readme
  https://twitter.com/browserslist


=> Started MongoDB.
autoprefixer: /home/MyDirectory/Rocket.Chat/app/theme/client/imports/general/base_old.css:158:2: start value has mixed support, consider using flex-start instead              autoprefixer: /home/MyDirectory/Rocket.Chat/app/theme/client/imports/general/base_old.css:158:2: start value has mixed support, consider using flex-start instead              W20190705-22:57:46.512(9)? (STDERR) /home/MyDirectory/.meteor/packages/meteor-tool/.1.8.1.ani1yi.p0f9s++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:280
W20190705-22:57:46.532(9)? (STDERR)                                             throw(ex);
W20190705-22:57:46.532(9)? (STDERR)                                             ^
W20190705-22:57:46.532(9)? (STDERR)
W20190705-22:57:46.533(9)? (STDERR) TypeError: Cannot read property 'constructor' of undefined
W20190705-22:57:46.533(9)? (STDERR)     at exposeSubscription (packages/meteorhacks_meteorx.js:61:41)
W20190705-22:57:46.533(9)? (STDERR)     at exposeLivedata (packages/meteorhacks_meteorx.js:31:5)
W20190705-22:57:46.533(9)? (STDERR)     at packages/meteorhacks_meteorx.js:175:1
W20190705-22:57:46.533(9)? (STDERR)     at packages/meteorhacks_meteorx.js:200:4
W20190705-22:57:46.533(9)? (STDERR)     at packages/meteorhacks_meteorx.js:208:3
W20190705-22:57:46.533(9)? (STDERR)     at /home/MyDirectory/Rocket.Chat/.meteor/local/build/programs/server/boot.js:419:36
W20190705-22:57:46.533(9)? (STDERR)     at Array.forEach (<anonymous>)
W20190705-22:57:46.533(9)? (STDERR)     at /home/MyDirectory/Rocket.Chat/.meteor/local/build/programs/server/boot.js:228:19
W20190705-22:57:46.533(9)? (STDERR)     at /home/MyDirectory/Rocket.Chat/.meteor/local/build/programs/server/boot.js:479:5
W20190705-22:57:46.534(9)? (STDERR)     at Function.run (/home/MyDirectory/Rocket.Chat/.meteor/local/build/programs/server/profile.js:510:12)
W20190705-22:57:46.534(9)? (STDERR)     at /home/MyDirectory/Rocket.Chat/.meteor/local/build/programs/server/boot.js:478:11
=> Exited with code: 1`



What am I doing wrong, and how can I fix this?

**Comment by semartsal** [comment]
> I apologize, I clearly missed where to find this feature, super ! Perhaps other people not finding it :
> pruning_per_channel

hi, anybody find this features ? I need to auto clear channel message on last 3-5 days
can't find this in 2.1.0 release (

**Comment by reetp** [comment]
> hi, anybody find this features ? I need to auto clear channel message on last 3-5 days
> can't find this in 2.1.0 release (

I don't believe you can do that. You can only clear messages 'older than'

You can open a NFR if you want:
https://github.com/RocketChat/feature-requests

**Comment by semartsal** [comment]
> > hi, anybody find this features ? I need to auto clear channel message on last 3-5 days
> > can't find this in 2.1.0 release (
> 
> I don't believe you can do that. You can only clear messages 'older than'
> 
> You can open a NFR if you want:
> https://github.com/RocketChat/feature-requests

I need AUTO delete message older then 5 day for expamle, for some channel, not for all.
and on the screen exactly what I need, but a can't find this in channel prune properties
only delete "newer" or "older".
2.1.0 release include this prune property ?
thnx for any answer

**Comment by reetp** [comment]
Please read this.
https://rocket.chat/docs/administrator-guides/retention-policies/

If you can't find what you need then open a feature request. Just add a reference to this PR.

Obviously a PR with your code to add new functions is always welcome.

**Comment by semartsal** [comment]
> I apologize, I clearly missed where to find this feature, super ! Perhaps other people not finding it :
> pruning_per_channel

i found this. 
enable RP first, then go in bottom of Room Info. profit
tnx for this features!

**Comment by Gummikavalier** [comment]
A question regarding completeness of pruning functionality:

Does this also prune user data files created by `Export Messages` feature currently available for admins in RC 3.7.1?

The export feature offers the messages and attachments as a zip file by a direct link to the server.
For example: `https://example.com/ufs/GridFS:UserDataFiles/cge2fwH2UFHJv12D33/2020-10-21-username-data.zip`
